recipes/LibriSpeech/ASR/CTC/train_with_wav2vec_k2.py

```diff
@@ -0,0 +1,656 @@
+            )
+            raise e
+        logger.info(f"Writing {out_fst_path}")
+        with open(out_fst_path, "w") as f:
+            f.write(s)
+
+    arpa_path = arpa_dir / trigram_arpa_name
+    fst_path = output_dir / os.path.basename(trigram_fst_output_name)
+    _arpa_to_fst_single(arpa_path, fst_path, max_order=3)
+    if convert_4gram:
+        arpa_path = arpa_dir / fourgram_arpa_name
+        fst_path = output_dir / os.path.basename(fourgram_fst_output_name)
+        _arpa_to_fst_single(arpa_path, fst_path, max_order=4)
+
+
+if __name__ == "__main__":
+    # CLI:
+    hparams_file, run_opts, overrides = sb.parse_arguments(sys.argv[1:])
+
+    # If distributed_launch=True then
+    # create ddp_group with the right communication protocol
+    sb.utils.distributed.ddp_init_group(run_opts)
+
+    with open(hparams_file) as fin:
+        hparams = load_hyperpyyaml(fin, overrides)
+
+    # env_corrupt is not supported with k2 yet
+    if hparams.get("env_corrupt", None):
+        raise NotImplementedError("env_corrupt is not supported with k2 yet")
+
+    # Create experiment directory
+    sb.create_experiment_directory(
+        experiment_directory=hparams["output_folder"],
+        hyperparams_to_save=hparams_file,
+        overrides=overrides,
+    )
+
+    # Dataset prep (parsing Librispeech)
+    from librispeech_prepare import prepare_librispeech  # noqa
+
+    # multi-gpu (ddp) save data preparation
+    run_on_main(
+        prepare_librispeech,
+        kwargs={
+            "data_folder": hparams["data_folder"],
+            "tr_splits": hparams["train_splits"],
+            "dev_splits": hparams["dev_splits"],
+            "te_splits": hparams["test_splits"],
+            "save_folder": hparams["output_folder"],
+            "merge_lst": hparams["train_splits"],
+            "merge_name": "train.csv",
+            "skip_prep": hparams["skip_prep"],
+        },
+    )
+
+    # here we create the datasets objects as well as tokenization and encoding
+    train_data, valid_data, test_datasets = dataio_prepare(hparams)
+
+    # Create the lexicon.txt for k2 training
+    run_on_main(
+        get_lexicon,
+        kwargs={
+            "lang_dir": hparams["lang_dir"],
+            "csv_files": [hparams["output_folder"] + "/train.csv"],
+            "extra_vocab_files": [hparams["vocab_file"]],
+            "add_word_boundary": hparams["add_word_boundary"],
+        },
+    )
+
+    # Create the lang directory for k2 training
+    run_on_main(
+        prepare_lang,
+        kwargs={
+            "lang_dir": hparams["lang_dir"],
+            "sil_prob": hparams["sil_prob"],
+        },
+    )
+
+    lexicon = Lexicon(hparams["lang_dir"])
+
+    # Trainer initialization
+    asr_brain = ASR(
+        modules=hparams["modules"],
+        hparams=hparams,
+        run_opts=run_opts,
+        checkpointer=hparams["checkpointer"],
+    )
+
+    need_G = hparams.get("use_HLG", False) in [True, "True"]
```

Comment:
Can we have the ontology bellow encapsulated in a proper function with a proper docstring explaining what is going on? We should avoid overloading the main function.

Reply:
Done